shared/api/v3/course.ts

```diff
@@ -0,0 +1,160 @@
+            const endTime = stringifyTime(schedule.endTime);
+            return schedule.locations.map((location) => ({
+                scheduleDays: days,
+                scheduleStartTime: startTime,
+                scheduleEndTime: endTime,
+                scheduleStartDate: startDate,
+                scheduleEndDate: endDate,
+                scheduleTermCount: termCount,
+                scheduleTerms: terms,
+                scheduleLocation: location,
+            }));
+        }),
+        courseCredits: s.credits,
+        courseSeatsTotal: s.seatsTotal,
+        courseSeatsFilled: s.seatsFilled,
+        courseWaitlistLength: null,
+        courseEnrollmentStatus: stringifySectionStatusEnum(s.status),
+        permCount: s.permCount,
+    };
+}
+
+export function courseListFromV4SectionList(sections: APIv4.Section[]): {
+    terms: [];
+    courses: Course[];
+} {
+    return {
+        terms: [],
+        courses: sections.map(courseFromV4Section),
+    };
+}
+
+function stringifyWeekdayEnumArray(arr: APIv4.WeekdayEnum[]): string {
+    return arr.join("");
+}
+
+function stringifyTime(secs: integer): string {
+    if (secs < 0 || secs >= 24 * 60 * 60) {
+        throw Error(`Time out of range: ${secs}`);
+    }
+    const total_minutes = secs / 60;
+
+    if (!Number.isInteger(total_minutes)) {
+        throw Error(`Time too precise to fit in hours:minutes format: ${secs}`);
+    }
+
+    const hours = Math.floor(total_minutes / 60);
+    const minutes = total_minutes % 60;
+
+    const hours_string = hours.toString().padStart(2, "0");
+    const minutes_string = minutes.toString().padStart(2, "0");
+
+    return `${hours_string}:${minutes_string}`;
+}
+
+function stringifyCourseDate(date: APIv4.CourseDate): string {
+    const padded_month = date.month.toString().padStart(2, "0");
+    const padded_day = date.day.toString().padStart(2, "0");
+    return `${date.year}-${padded_month}-${padded_day}`;
+}
+
+function stringifySectionTermCode(identifier: APIv4.SectionIdentifier): string {
+    const half = identifier.half;
+    const halfString = half !== null ? `${half.prefix}${half.number}` : "";
+    return `${identifier.term}${identifier.year}${halfString}`;
+}
+
+function getSectionTermFractions(identifier: APIv4.SectionIdentifier): {
+    termCount: number;
+    terms: number[];
+} {
+    const half = identifier.half;
+    if (half === null) {
+        return {
+            termCount: 1,
+            terms: [0],
+        };
+    } else if (half.prefix === "F" || half.prefix === "P") {
+        if (half.number > 2) {
+            throw Error(
+                `Fall/Spring semester divided into too many parts: ${half.number}`,
+            );
+        }
+        return {
+            termCount: 2,
+            terms: [half.number - 1],
+        };
+    } else {
+        // Best guess given available data
+        return {
+            termCount: half.number,
+            terms: [half.number - 1],
+        };
+    }
```

Comment:
so, here's a style comment that maybe deviates from my (otherwise religious) devotion to FP-style conventions. when `if` blocks return early, don't put the remaining code inside `else` blocks:
```typescript
if (half === null) {
  return thing1;
} else if (half.prefix /* ... */) {
  return thing2;
} else {
  return thing3;
}

// do
if (half === null) {
  return thing1;
}
if (half.prefix /* ... */) {
  return thing2;
}
return thing3;
```

there are two (subjective) advantages to this--
- reduces nesting.
- makes it clearer (? depends on who you ask, i guess) what branches of the logic are "short-circuit"/special cases, and which path corresponds to the "default"/catch-all path. though to be pedantic, in practice the "default" path here should be hit rarely/almost-never given the data we have/are working with, so application of this rule to this specific case may be taken with a grain of salt. my point remains as a general rule of thumb.

just to claim that this rule isn't _purely_ my own preference, but something that various other devs in the world online also care about-- it's sometimes called the "golden margin" rule, or the "align the happy path to the left edge" rule, articles/videos on those here:
- https://www.youtube.com/watch?v=CFRhGnuXG-4&t=216s
- https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

Reply:
I am aware of this style I definitely agree with it. But like you said, the `else` case here is definitely a more anomalous situation, and doesn't represent the Happy Path, which is why I formatted it like this. I suppose I could change the conditions so that the Happy Path goes into the `else` case, and then I could take it out of the `if` altogether.

Reply:
it's up to you then, feel free to keep as it